pkg/dao/crd/dao.go

```diff
@@ -0,0 +1,306 @@
+//
+// Copyright (c) 2018 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package dao
+
+import (
+	"fmt"
+
+	automationbrokerv1 "github.com/automationbroker/broker-client-go/client/clientset/versioned/typed/automationbroker.io/v1"
+	v1 "github.com/automationbroker/broker-client-go/pkg/apis/automationbroker.io/v1"
+	"github.com/openshift/ansible-service-broker/pkg/apb"
+	"github.com/openshift/ansible-service-broker/pkg/clients"
+	logutil "github.com/openshift/ansible-service-broker/pkg/util/logging"
+	"github.com/pborman/uuid"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var log = logutil.NewLog()
+
+const (
+	// instanceLabel for the job state to track which instance created it.
+	jobStateInstanceLabel string = "instanceId"
+	jobStateLabel         string = "state"
+)
+
+// Dao - object to interface with the data store.
+type Dao struct {
+	client    automationbrokerv1.AutomationbrokerV1Interface
+	namespace string
+}
+
+// NewDao - Create a new Dao object
+func NewDao(namespace string) (*Dao, error) {
+	dao := Dao{namespace: namespace}
+
+	crdClient, err := clients.CRDClient()
+	if err != nil {
+		return nil, err
+	}
+	dao.client = crdClient.AutomationbrokerV1()
+	return &dao, nil
+}
+
+// GetSpec - Retrieve the spec from the k8s API.
+func (d *Dao) GetSpec(id string) (*apb.Spec, error) {
+	log.Debugf("get spec: %v", id)
+	s, err := d.client.Bundles(d.namespace).Get(id, metav1.GetOptions{})
+	if err != nil {
+		log.Errorf("unable to get bundle from k8s api - %v", err)
+		return nil, err
+	}
+	return bundleToSpec(s.Spec, s.GetName()), nil
```

Comment:
So it's expected that the dao layer will need to transform between the ASB types and their CRD counterparts for most of these, so the rest of the codebase is getting objects they expect?

Reply:
Yes, that is the plan and what I was hoping to do here.